### PR TITLE
kernel: Add "per thread" timeslice mechanism

### PR DESCRIPTION
**Pull request by andyross**
Zephyr's timeslice implementation has always been somewhat primitive.
You get a global timeslice that applies broadly to the whole bottom of
the priority space, with no ability (beyond that one priority
threshold) to tune it to work on certain threads, etc...

This adds an (optionally configurable) API that allows timeslicing to
be controlled on a per-thread basis: any thread at any priority can be
set to timeslice, for a configurable per-thread slice time, and at the
end of its slice a callback can be provided that can take action.
This allows the application to implement things like responsiveness
heuristics, "fair" scheduling algorithms, etc... without requiring
that facility in the core kernel.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

### CODE REVIEW

**Comment by andyross** [comment]
Here for API & mechanism review.  Not quite ready for merge yet as I don't have a proper test written yet.

**Review by peter-mitsis — DISMISSED** [review]

**Review by dcpleung — DISMISSED** [review]

**Comment by mwasko** [comment]
@andyross I have two questions:

1. If the timesliced thread time elapse and it remain as a highest priority thread then I assume it will be re-scheduled again immediately with the same timeslice for processing?

2. If the timesliced high priority thread will be set to medium priority on callback then will it be but at the end of scheduling queue (assuming that there are other medium priority tasks waiting for processing)?

**Review by mwasko — DISMISSED** [review]

**Review by mwasko — COMMENTED** [review]
@andyross is it possible to add optional 'reset' of thread time_slice cycles when thread time slice is updated by k_thread_time_slice_set?

So in situation when thread is in the middle of processing (with some initial time_slice value) and we set new time_slice using k_thread_time_slice_set(), we would like thread to be allowed to continue processing for entire new time_slice value.

**Comment by andyross** [comment]
That's a reasonable idea.  It's a little complicated to implement with the current scheme because the active cycle count is stored on the CPU struct and not the thread; we'd have to move it (at the cost of an extra word of storage in the thread struct) or else write a locked loop over the active cores to find the thread (potentially a latency issue for systems with many cpus).

Given this hardware, I don't see why the first wouldn't be acceptable.  Let me think on it a bit before promising? :)

**Comment by mwasko** [comment]
@andyross when calculating thread processing time, does Zephyr takes correction for time consumed on context switching?

**Comment by mwasko** [comment]
@andyross you mentioned in description that thread at any priority can be set to timeslice, does it mean that it can also be applied to non-preemptive (negative priority) threads?

**Review by carlescufi — DISMISSED** [review]

**Comment by andyross** [comment]
> when calculating thread processing time, does Zephyr takes correction for time consumed on context switching?

It does not.  The scales are too different to matter.  Timeslicing is done in units of ticks (~100's of us) where context switching is at most a few hundred CPU cycles (100's of ns).

> does it mean that it can also be applied to non-preemptive (negative priority) threads?

It does, because it's symmetric and there's no reason not to provide this as a feature.  Just in general though it's probably a bad idea to mix cooperative thread scheduling (which is really a synchronization paradigm, not a scheduling optimization!) with the idea of a thread being preempted.

**Comment by andyross** [comment]
Dust this off and add a test.  Seems to work for me, modulo known timer sensitivity in the schedule_api test where I chose to put the case.  I don't **think** this made anything worse, but let's see what CI says.

**Comment by mwasko** [comment]
@andyross one more thing to consider. Some threads might implement atomic operations that will disable interrupts for certain processing blocks and in result they potentially can consume more time then configured time slice.

I would like to suggest to perform a check whether such thread time slice violation occurred and report warning message. It could also be helpful to propagate the violation information to the slice expiration callback.

**Comment by andyross** [comment]
@mwasko That's something that could easily be done inside the callback; I'm not sure there's value in providing that as an OS feature.  In fact decisionmaking like that was what the callback was intended to enable.

Also: my gut turns out to have been right, and there is a regression hidden here.  One of the last minute fixups here had to swap the ordering of the "set current thread" and "reset time slice" steps (because now resetting the slice time depends on what thread is running).  In theory that shouldn't change anything, but in practice lots of scheduler behavior (and the schedule_api test in particular) tends to be extremely sensitive to timing deltas like this.  Will figure it out.

**Comment by andyross** [comment]
One more spin, cleaning up an induced bug with the legacy timeout API.  Also: see the doxygen comments for a clarification about an important API detail regarding tick timing.  Re-review (at least a little) please.

**Comment by andyross** [comment]
Grr... one more.  qemu_cortex_m3 isn't built by default twister but is by CI, and its use of multiq by default showed up a filter I'd miscopied in the new test YAML.

**Review by peter-mitsis — COMMENTED** [review]
Two items on a theme.
  1. There seems to be a mixing of 'int32_t' and 'int' types with respect to the time slicing.
  2. Since the unit of measurement is ticks, why not use 'k_ticks_t'?

**Comment by andyross** [comment]
1. The "int" slice time was pre-existing, I left the type as it was during the modification out of conservatism, but there's no particular reason it shouldn't be changed here.

2. k_ticks_t is routinely a 64 bit quantity on 32 bit systems, so trying to use it would have code size and performance impact that we don't want (no one needs a 4B+ tick timeslice, obviously).

File: include/kernel.h [hunk 1]
```diff
@@ -864,6 +864,47 @@ __syscall void k_thread_resume(k_tid_t thread);
...
+ * not yet reflect the expired ticks.  Applications wishing to make
+ * fine-grained timing decisions within this callback should use the
+ * cycle API, or derived facilities like k_thread_runtime_stats_get().
+ *
+ * @param th A valid, initialized thread
+ * @param slice_ticks Maximum timeslice, in ticks
+ * @param expired Callback function called on slice expiration
+ * @param data Parameter for the expiration handler
+ */
+void k_thread_time_slice_set(struct k_thread *th, int32_t slice_ticks,
```
Comment (ceolin) [inline include/kernel.h:905]: I think it should receive `k_ticks_t slice_ticks`, and then if you need you convert internally.

File: kernel/include/ksched.h [hunk 2]
```diff
@@ -56,7 +56,7 @@ bool z_set_prio(struct k_thread *thread, int prio);
 void *z_get_next_switch_handle(void *interrupted);
 void idle(void *unused1, void *unused2, void *unused3);
 void z_time_slice(int ticks);
-void z_reset_time_slice(void);
+void z_reset_time_slice(struct k_thread *curr);
```
Comment (ceolin) [inline kernel/include/ksched.h:59]: Does it really need to change ? This API always end up receiving the current thread AFAIU.

File: include/kernel.h [hunk 3]
```diff
@@ -864,6 +864,47 @@ __syscall void k_thread_resume(k_tid_t thread);
...
+			     k_thread_timeslice_fn_t expired, void *data);
```
Comment (ceolin) [inline include/kernel.h:906]: maybe `const void *data` ? Since it is a caller data pointer ?

**Review by ceolin — APPROVED** [review]
I added minor comments, the behavior and implementation looks good for me.

**Review by peter-mitsis — APPROVED** [review]

**Pull request merged**